### ISSUE TITLE
Do not ignore index when passing a Series to a parameter that accepts flexible types

### ISSUE THREAD

**Issue by ResidentMario** [ResidentMario/geoplot#160]
Currently when you pass a `Series` to a flexible-type parameter like `hue` or `scale`, the `Series` index will be ignored and the `Series` will be treated the same way as a flat `numpy` `array` or a `list`: as a container whose mapping onto the data is determined by the indicial order of its data items.

`geoplot` should instead match the `Series` index against the index of the input `GeoDataFrame`. If the `Series` contains the same items present in the `GeoDataFrame` (in any order), use the index match. If it contains all of the same items present in the `GeoDataFrame`, plus additional items, raise a user warning but continue with plotting. If it does not contain all of the expected items, raise an error.

**Comment by deppen8** [comment]
Hi @ResidentMario, been a fan of `geoplot` for a while so thought I would give this a shot. The code below is a generic function that should do the validation you outlined plus some `pytest` tests that make sure it works properly.

I wasn't sure exactly where to put it within `geoplot`, so I didn't make a PR yet, but if you want to give me a bit of direction, I can add it in properly. Otherwise, feel free to use this code yourself.

Some notes:
- It doesn't do the final step of aligning the indices of the Series and (Geo)DataFrame. I figure that will be done if the `validate` function returns `True`.
- I chose to raise an error if either the Series or the DataFrame indices had duplicates. This simplifies the rest of the logic a great deal.
- The function could be simplified to return `False` instead of different errors, but I chose to raise errors to because it could fail for a variety of reasons. 
- I used plain DataFrames for the tests because they were simpler to specify, but an improved version would probably use proper GeoDataFrames instead.

Let me know what you think.

```
import warnings
import pandas as pd

def validate_series_in_df(s: pd.Series, df: pd.DataFrame) -> bool:
    # validate unique indices
    if any(s.index.duplicated()):
        raise ValueError("Series index contains duplicates")
    if any(df.index.duplicated()):
        raise ValueError("DataFrame index contains duplicates")

    if len(s.index) == len(df.index):
        if s.index.equals(df.index):       # 1. same length, identical
            return True
        else:                               # 2. same length, mismatched
            raise ValueError("Not all values in DataFrame index have corresponding value in Series index")
    elif len(s.index) > len(df.index):
        if all(df.index.isin(s.index)):    # 3. longer series, all df items have a match
            warnings.warn("Extra items in Series that are not present in DataFrame, but continuing with plot.")
            return True
        else:                               # 4. longer series, not all df items have a match
            raise ValueError("Not all values in DataFrame index have corresponding value in Series index")
    elif len(s.index) < len(df.index):     # 5. shorter series
        raise ValueError("Not all values in DataFrame index have corresponding value in Series index")


#########
# TESTS #
#########
import pytest

def test_series_duplicate_ix():
    s = pd.Series(["a", "b", "c", "d"], index=[0, 1, 1, 2])
    df = pd.DataFrame({'name':['foo', 'bar', 'baz'], 'end':["x", "y", "z"]}, index=[0, 1, 2])

    with pytest.raises(ValueError, match="Series index contains duplicates"):
        validate_series_in_df(s, df)


def test_df_duplicate_ix():
    s = pd.Series(["a", "b", "c"], index=[0, 1, 2])
    df = pd.DataFrame({'name':['foo', 'bar', 'baz', 'bop'], 'end':["w", "x", "y", "z"]}, index=[0, 1, 1, 2])

    with pytest.raises(ValueError, match="DataFrame index contains duplicates"):
        validate_series_in_df(s, df)


def test_equal_len_equal_ix():
    s = pd.Series(["a", "b", "c"], index=[0, 1, 2])
    df = pd.DataFrame({'name':['foo', 'bar', 'baz'], 'end':["x", "y", "z"]}, index=[0, 1, 2])

    assert validate_series_in_df(s, df) == True


def test_equal_len_diff_ix():
    s = pd.Series(["a", "b", "c"], index=[0, 1, 9])
    df = pd.DataFrame({'name':['foo', 'bar', 'baz'], 'end':["x", "y", "z"]}, index=[0, 1, 2])

    with pytest.raises(ValueError, match="Not all values in DataFrame index have corresponding value in Series index"):
        validate_series_in_df(s, df)

@pytest.mark.filterwarnings("ignore:Extra items")
def test_long_s_all_df_match():
    s = pd.Series(["a", "b", "c", "d"], index=[0, 1, 2, 3])
    df = pd.DataFrame({'name':['foo', 'bar', 'baz'], 'end':["x", "y", "z"]}, index=[0, 1, 2])

    assert validate_series_in_df(s, df) == True

def test_long_s_all_df_match_warning():
    s = pd.Series(["a", "b", "c", "d"], index=[0, 1, 2, 3])
    df = pd.DataFrame({'name':['foo', 'bar', 'baz'], 'end':["x", "y", "z"]}, index=[0, 1, 2])

    with pytest.warns(Warning) as record:
        validate_series_in_df(s, df)
    
    assert str(record[0].message) == "Extra items in Series that are not present in DataFrame, but continuing with plot."    


def test_long_s_not_all_df_match():
    s = pd.Series(["a", "b", "c", "d"], index=[0, 1, 2, 9])
    df = pd.DataFrame({'name':['foo', 'bar', 'baz'], 'end':["x", "y", "z"]}, index=[0, 1, 3])

    with pytest.raises(ValueError, match="Not all values in DataFrame index have corresponding value in Series index"):
        validate_series_in_df(s, df)
    

def test_short_s():
    s = pd.Series(["a", "b", "c"], index=[0, 1, 2])
    df = pd.DataFrame({'name':['foo', 'bar', 'baz', 'bop'], 'end':["w", "x", "y", "z"]}, index=[0, 1, 2, 3])

    with pytest.raises(ValueError, match="Not all values in DataFrame index have corresponding value in Series index"):
        validate_series_in_df(s, df)
```

**Comment by ResidentMario** [comment]
Thanks for the contribution! It would be a lot easier for me to review this code if you could go ahead and open up a pull request against `geoplot` with it.

In terms of where the code should live, the input validation tests seem like they could go in [`input_tests.py`](https://github.com/ResidentMario/geoplot/blob/master/tests/input_tests.py). The method itself could go in the main [`geoplot.py`](https://github.com/ResidentMario/geoplot/blob/master/geoplot/geoplot.py) file, which contains a section for "Helper methods".

**Comment by ResidentMario** [comment]
This has been bothering me for a while, so I went ahead and implemented it myself in #165.

@deppen8 having done the full implementation, some comments on your code:

* Restricting the input `GeoDataFrame` to have a fully unique index is unnecessarily restrictive. There are reasons why a user might happen to have a non-unique index, and the correct behavior for this case is well-defined (see `pd.Series.reindex`), so there's no reason we can't support that.
* Restricting the input `Series` to have the same length as the input `GeoDataFrame` is also unnecessarily restrictive. It's fine if the index values in the `Series` are a superset of those in the `GeoDataFrame`; the extra ones just won't be carried over.

You can see the implementation by visiting the PR. Closing this issue now.
